docs/quantum_gate.rst

```diff
@@ -1,4 +1,366 @@
+.. _quantum_gates:
+
 Quantum Gates
 =============
 
-contains gate details
\ No newline at end of file
+Gates in OpenQL are the constructs that refer to operations to be executed somehow on the computing platform.
+
+A gate refers to an operation and to zero or more operands.
+
+Gates are organized in circuits as vectors of gates, i.e. linear sequences of gates.
+A circuit defines the operation of a kernel.
+And a program consists of multiple kernels.
+
+Gates can be subdivided in several kinds.
+This is useful in the description of the passes below.
+
+
+First, gates can be subdivided according to where their execution has effect:
+
+- quantum gates; these gates execute in the quantum computing hardware; these gates have at least one qubit as (implicit or explicit) operand; these gates can have classical registers as operand as well and may rely on some execution capability in classical hardware
+
+- classical gates; these gates execute in classical hardware only; these don't have any qubit as operand, only zero or more classical registers
+
+- directives; these gates execute neither in quantum nor in classical hardware; these look like gates but don't influence execution, e.g. the display gate
+
+
+Quantum gates can be subdivided in several ways from the number of operands they take; this becomes relevant when gates are mapped on the quantum computing platform:
+
+- one-qubit gates; quantum gates operating on one qubit
+
+- two-qubit gates; quantum gates operating on two qubits;
+  these are the main objective in mapping since two-qubit gates require their qubit operands to be connected in the hardware
+
+- multi-qubit gates; quantum gates operating (implicitly or explicitly) on more than two qubits;
+  these must be decomposed to one-qubit and two-qubit gates because more-qubit primitive gates are not supported by the quantum platform
+
+
+Quantum gates can also be subdivided seen from the state of a qubit:
+
+- preparation gates; (usually one-qubit) gates taking qubits in an undefined state and bringing them in a particular defined state
+
+- rotation gates; gates that perform unitary rotations on the state of the operand qubits;
+  examples are identity, x, rx(pi), cnot, swap, and toffoli.
+
+- measurement gates; gates that measure out the operand qubits, leaving them in a base state; the measurement result can be left in a classical register
+
+- scheduling gates; gates that only influence execution timing regarding the operand qubits;
+  they provide a cycle window for the qubit state to be operated upon before further use;
+  examples are the wait and barrier gates
+
+
+Particular classes of quantum gates can be further recognized:
```

Comment:
I am assuming that this subdivision has not consequence in the code as in there are no actual `classes` that link to this. 
Pauli gates (and some rotations) are also Cliffords. As such I am confused with how I should interpret this.

Reply:
The definitions help in the other parts of the documentation.
Yes, by Cliffords I meant those single-qubit rotations that form a group with 24 equivalence classes. There is an API to create one from this group. And now there is an optimizer that reduces sequences of these to the smallest sequence.